src/compiler/types.ts

```diff
@@ -5334,6 +5334,7 @@ namespace ts {
         fileExists?(path: string): boolean;
         readFile?(path: string): string | undefined;
         getSourceFiles?(): ReadonlyArray<SourceFile>; // Used for cached resolutions to find symlinks without traversing the fs (again)

```

Comment:
I think the API would be better if this and `getCommonSourceDirectory` weren't optional members of this interface, and were just passed in as values in the one place they're needed.

Reply:
Hmmmmm they're on the interface because that interface is how the declaration emitter shuffles bits of its emit host into the checker to fulfill these needs. And I don't _particularly_ want to pass in a second host object with the kinda-nice-to-have-for-one-of-the-entry-points methods (especially since `{moduleResolverHost: host, declarationModuleResolverHost: host}` in `declarations.ts` would look like we were going mad).

Reply:
I didn't mean to add a second host for those methods, but that a host isn't necessary at all. Both methods are immediately called in the first function and not used by anything else in `moduleSpecifiers`.
Making the method optional seems to cover up problems, such as the fact that `getCommonSourceDirectory` isn't defined on `TypeCheckerHost` (although it might happen to be defined at runtime anyway, though possibly with a different type), and users of other functions will think they need to provide these methods even when they're unused outside of `getModuleSpecifierForDeclarationFile`. It also looks like `host.getSourceFiles()` is non-optional so testing for its existence is dead code.

Reply:
A host is merely a group of argument methods to delegate to, no? If I need two, it's reasonable to group them into a second host object.